### PR TITLE
Add s3 support

### PR DESCRIPTION
**Pull request by rgooler**
Since s3 support is available in ansible, and its a semi-common way to store archived files like .wars, I figured s3 support would be a good thing. And it keeps me from having to maintain a fork. :)

Tested and working. Also had some issues with installing on ubuntu, as the directories for wars were not created properly.

### CODE REVIEW

**Comment by robertdebock** [comment]
I agree that support for s3 is a great plan. Let me review your code and probably ask for a few questions.

File: tasks/tomcat_wars.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+---
+- name: "Ensure webapp dir exists (url)"
+  shell: "mkdir -p {{ item.directory | default(tomcat_directory) }}/{{ tomcat_archive[item.version | default (tomcat_version)]['dir'] }}/webapps"
```
Comment (robertdebock) [inline tasks/tomcat_wars.yml]: Using the module `file` would be better.

File: tasks/tomcat_wars.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  when:
+    - "tomcat_wars.url is defined"
+  with_items: "{{ tomcat_wars.url }}"
+- name: "loop over tomcat_wars"
```
Comment (robertdebock) [inline tasks/tomcat_wars.yml]: Empty lines between tasks would improve readability.

File: tasks/tomcat_wars.yml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  get_url:
+    dest: "{{ item.directory | default(tomcat_directory) }}/{{ tomcat_archive[item.version | default (tomcat_version)]['dir'] }}/webapps/"
+    url: "{{ item }}"
+    validate_certs: false
+  when:
+    - "tomcat_wars.url is defined"
+  with_items: "{{ tomcat_wars.url }}"
+- name: "Ensure webapp dir exists (s3)"
+  shell: "mkdir -p {{ item.directory | default(tomcat_directory) }}/{{ tomcat_archive[item.version | default (tomcat_version)]['dir'] }}/webapps"
```
Comment (robertdebock) [inline tasks/tomcat_wars.yml]: Again: the module `file` would be better. Just to explain; shell is `changed` all the time, while `file` is idempotent.

File: tasks/tomcat_wars.yml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  when:
+    - "tomcat_wars.s3 is defined"
+  with_items: "{{ tomcat_wars.s3 }}"
+- name: "loop over tomcat_wars in s3"
+  aws_s3:
```
Comment (robertdebock) [inline tasks/tomcat_wars.yml]: Great module! I think it's probably easier to refer to the s3 object in variables, something like this:
```
tomcat_layout:
  - name: appone
    wars:
      - url: https://tomcat.apache.org/tomcat-6.0-doc/appdev/sample/sample.war
      - s3:
        - bucket: mybucket
          object: /my/desired/key.txt
```

(I'm not sure about the suggested layout, seems a bit complex...) I'm open to suggestions.

Comment (robertdebock) [inline tasks/tomcat_wars.yml]: By the way: If the bucket is publicly available, the `url:` could be used to deploy the war right?

File: tasks/tomcat_wars.yml (hunk 2 above)
Comment (rgooler) [reply tasks/tomcat_wars.yml]: Yeah, will fix. Had issues with molecule until I learned to use it.

File: tasks/tomcat_wars.yml (hunk 3 above)
Comment (rgooler) [reply tasks/tomcat_wars.yml]: mkdir -p is idempotent, but I'll change it. :)

File: tasks/tomcat_wars.yml (hunk 4 above)
Comment (rgooler) [reply tasks/tomcat_wars.yml]: Thats the exact layout I built for this. :)

Comment (rgooler) [reply tasks/tomcat_wars.yml]: Yes. But this lets you deploy out of private s3 buckets, and not make your war files readable to the world. Pretty important when your war contains sensitive or proprietary data.

**Comment by rgooler** [comment]
Currently seems to not be starting my app up locally - figuring out why it seems to be installing the sample app instead of the one I gave it.

**Comment by rgooler** [comment]
This is failing on ansible 2.3 and I don't understand why. Fixed the rest of the tests, though.

File: molecule/ubuntu-16.04/playbook.yml [hunk 5]
```diff
@@ -22,3 +22,12 @@
           ajp_port: 8009
           wars:
             - url: https://tomcat.apache.org/tomcat-6.0-doc/appdev/sample/sample.war
+          roles:
```
Comment (robertdebock) [inline molecule/ubuntu-16.04/playbook.yml]: This is nice stuff. Good suggestion, well defined in these variables.

I'll see if I can merge just these changes, if that's not possible I'll ask you to make another merge request just for this feature.

Comment (rgooler) [reply molecule/ubuntu-16.04/playbook.yml]: I actually did this cause I'm trying to debug the server I'm building to figure out why tomcat is starting, but my app doesn't seem to be. Eventually gonna wire up LDAP in there, once I figure out how.

**Comment by rgooler** [comment]
Ok, now everything works - optionally clears out tomcat and runs JUST the war you choose hosted at / if you only pass in one war file to be installed, and allows you to add users and roles to tomcat-users.xml

Only thing that doesnt work is ansible 2.3, and that appears to be because 2.3 doesn't support using "when:" on actions.

**Comment by robertdebock** [comment]
Closing this request, because of conflict in merging and s3 can also be supported using the (existing) url structure:
```yaml
---
tomcat_instances:
  - name: "{{ tomcat_name }}"
    url:
      - https://example.com/some.war
```

**Pull request closed without merging**